columnflow/columnar_util.py

```diff
@@ -1600,6 +1614,35 @@ def _get_produced_columns(self, call_cache: set | None = None) -> set[str]:
     def produced_columns(self) -> set[str]:
         return self._get_produced_columns()
 
+    def _check_columns(self, ak_array: ak.Array, io_flag: IOFlag):
+        """
+        Check if awkward array contains at least one column matching each
+        entry in 'uses' or 'produces' and raise Exception if none were found.
+        """
+
+        columns = self._get_columns(
+            io_flag=io_flag,
+            # only check own columns
+            no_dependencies=True,
+        )
+        action = (
+            "receive" if io_flag == self.IOFlag.USES else
+            "produce" if io_flag == self.IOFlag.PRODUCES else
+            "find"
+        )
+
+        missing = set()
+        for column in columns:
+            found = ak_array.layout.form.select_columns(column).columns()
```

Comment:
Maybe one can use the `has_ak_column` helper here? This would also make sense as `column` could also be a `Route` object.

Reply:
I see the point about `column` potentially being a `Route`, but looking at the code again it looks like the only way this can happen is if the columns are given explicitly as routes in the array function decorator, i.e. `uses={Route("foo")}`. As far as I know, in the current code base only strings are used, so do we want to support `Route` objects explicitly?

In any case, I think we need `select_columns` here because that function automatically expands wildcards, while with `has_ak_columns` one would need the actual column names. For `Route` objects, it should be straightforward to just pass `column.string_column` to `select_columns` instead.

Reply:
Yup, makes sense :+1: